### PR TITLE
0.8 port

### PR DESCRIPTION
**Pull request by atotic**

### CODE REVIEW

File: demo.html [hunk 1]
```diff
@@ -4,10 +4,15 @@
 <head>
   <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes">
   <title>Google Map demo</title>
-  <script src="../webcomponentsjs/webcomponents.min.js"></script>
+  <script src="../webcomponentsjs/webcomponents.js"></script>
```
Comment (ebidel) [inline demo.html]: I prefer the min files because it sets a good example.

File: demo.html [hunk 2]
```diff
@@ -4,10 +4,15 @@
...
   <link rel="import" href="google-map.html">
+  <link rel="import" href="google-map-marker.html">
```
Comment (ebidel) [inline demo.html]: Is there a reason you broke this out? google-map-marker.html is only useful as a distributed node to google-map. I still think keeping them together makes sense.

File: google-map-directions.html [hunk 3]
```diff
@@ -53,8 +58,10 @@
        * @type google.maps.Map
        * @default null
        */
-      map: null,
-
+      map: {
+        type: Object,
+        observer: 'mapChanged'
```
Comment (ebidel) [inline google-map-directions.html]: Should we set some default values on these properties. The doc strings still say `@default null`

File: google-map-directions.html [hunk 4]
```diff
@@ -80,7 +87,10 @@
        * @type string
```
Comment (ebidel) [inline google-map-directions.html]: `@type String`

File: google-map-directions.html [hunk 5]
```diff
@@ -71,7 +78,7 @@
        * @type string|google.maps.LatLng
```
Comment (ebidel) [inline google-map-directions.html]: Please update the doc strings throughout

File: google-map-directions.html [hunk 6]
```diff
@@ -116,59 +129,60 @@
...
-          return;
-        }
-
-        // Construct a directionsService if necessary.
-        // Wait until here where the maps api has loaded and directions are actually needed.
-        if (!this.directionsService) {
-          this.directionsService = new google.maps.DirectionsService();
+      response: {
+        type: Object,
+        observer: 'responseChanged'
```
Comment (ebidel) [inline google-map-directions.html]: Should this `notify: true`?

File: google-map.html [hunk 7]
```diff
@@ -511,217 +147,295 @@
     }
 
   </style>
+  <template>
+    <google-maps-api 
+      api-key="{{apiKey}}" 
+      client-id="{{clientId}}" 
+      version="{{version}}" 
+      libraries="{{libraries}}"
```
Comment (ebidel) [inline google-map.html]: I think all of these can be one-way bindings

File: google-map.html [hunk 8]
```diff
@@ -511,217 +147,295 @@
...
+      /**
+       * A latitude to center the map on.
+       *
+       * @attribute latitude
+       * @type number
+       * @default 37.77493
+       */
+      latitude: {
+        type: Number,
+        value: 37.77493
```
Comment (ebidel) [inline google-map.html]: lat/lng should probably notify. I'd imagine people what to be notified on those. Also probably `relflectToAttribute`.

File: google-map.html [hunk 9]
```diff
@@ -511,217 +147,295 @@
...
-     */
-    mapType: 'roadmap', // roadmap, satellite, hybrid, terrain
+      /**
+       * When set, displays a map marker at the center point.
+       *
+       * @attribute showCenterMarker
+       * @type boolean
+       * @default false
+       */
+      showCenterMarker: {
```
Comment (ebidel) [inline google-map.html]: I'd actually like to deprecate this now that we have `<google-map-marker>`: https://github.com/GoogleWebComponents/google-map/issues/109

File: google-map.html [hunk 10]
```diff
@@ -511,217 +147,295 @@
...
+       * If set, custom styles can be applied to the map.
+       * For style documentation see developers.google.com/maps/documentation/javascript/reference#MapTypeStyle
+       *
+       * @attribute styles
+       * @type JSON encoded array
+       * @default null
+       */
+      styles: {
+        type: Object,
+        value: function() { return {}}
```
Comment (ebidel) [inline google-map.html]: spacing: `value: function() { return {}; }`

File: google-map.html [hunk 11]
```diff
@@ -732,6 +446,11 @@
       this.resize();
     },
 
+    detached: function() {
+      if (this._mutationObserver)
```
Comment (ebidel) [inline google-map.html]: place use brackets

File: google-map.html [hunk 12]
```diff
@@ -773,13 +492,20 @@
...
-          this.$.markers.getDistributedNodes());
+          Polymer.dom(this.$.markers).getDistributedNodes());
 
       if (this.centerMarker) {
         this.markers.push(this.centerMarker);
       }
 
-      this.onMutation(this, this.updateMarkers); // Watch for future updates.
+      // Watch for future updates.
+      if (this._mutationObserver)
```
Comment (ebidel) [inline google-map.html]: please use brackets

**Comment by ebidel** [comment]
Reviewed. Nice work. Needs a rebase.

File: demo.html (hunk 1 above)
Comment (atotic) [inline demo.html]: Changed to webcomponents-lite.js

File: demo.html (hunk 2 above)
Comment (atotic) [inline demo.html]: Several reasons:
- you might want to use google-map without markers, why download extra code
- 1 component per file is (mostly) the Polymer way
- google-map is 700+ lines, google-map-marker 400, having them both in same file made understanding harder

File: google-map-directions.html (hunk 3 above)
Comment (atotic) [inline google-map-directions.html]: Yes. Turns out that leaving values undefined makes computed properties not fire. I think specifying default value is a must. Done.

File: google-map-directions.html (hunk 5 above)
Comment (atotic) [inline google-map-directions.html]: I am going to wait on this until I upgrade the doc page to iron-doc. Hydrolysis has lots of new goodies, I think generating types automatically is one of them.

File: google-map-directions.html (hunk 6 above)
Comment (atotic) [inline google-map-directions.html]: sure. done.

File: google-map.html (hunk 7 above)
Comment (atotic) [inline google-map.html]: They can, I'll do it. The prevailing Polymer style right now is not to be too strict about this.

File: google-map.html (hunk 8 above)
Comment (atotic) [inline google-map.html]: doen.

File: google-map.html (hunk 9 above)
Comment (atotic) [inline google-map.html]: want to remove it?

File: google-map.html (hunk 10 above)
Comment (atotic) [inline google-map.html]: nit! Done.

File: google-map.html (hunk 11 above)
Comment (atotic) [inline google-map.html]: what is "place use brackets"

**Comment by atotic** [comment]
Might be nice work, but it no longer runs. This was my first port, must have rotted, will take a look tomorrow. I am getting:
UnauthorizedURLForClientIdMapError: https://developers.google.com/maps/documentation/business/clientside/auth#registering_authorized_urls
on http://localhost:8080/google-map/demo.html
which used to work.

Still needed:
- few comments above
- iron-doc port, and updating of comments
- old outstanding bugs:
marker removal tests fails, because getDistributedNodes() returns obsolete list. Should updateMarkers do querySelector() instead?
markers do not appear unless dom=shadow

File: google-map-directions.html (hunk 5 above)
Comment (ebidel) [inline google-map-directions.html]: Ack

File: google-map.html (hunk 9 above)
Comment (ebidel) [inline google-map.html]: Yes please :)

File: google-map.html (hunk 11 above)
Comment (ebidel) [inline google-map.html]: "please" use brackets. sorry

Comment (atotic) [inline google-map.html]: done

File: google-map.html (hunk 12 above)
Comment (atotic) [inline google-map.html]: done

File: google-map.html (hunk 9 above)
Comment (atotic) [inline google-map.html]: done

**Comment by ebidel** [comment]
LMK when I should take a look again

**Comment by atotic** [comment]
You can take a look now. 

Documentation is problematic, will need work. iron-doc-viewer is very different from core-component-viewer

**Comment by ebidel** [comment]
LGTM for now. Can you do a rebase/merge and I'll merge.

**Comment by atotic** [comment]
Just to make sure I am doing the right thing: (new to your workflow):
- rebase atotic/google-map#master to current GoogleWebComponents/google-map#master
- rebase/merge atotic/google-map#0.8-preview to new atotic/google-map#master

Correct?

**Comment by ebidel** [comment]
No sorry. We're not doing anything with master yet! I'm not sure why github can't merge this PR cleanly into the 0.8-preview branch.  I've merge these changes into 0.8-preview: https://github.com/GoogleWebComponents/google-map/commit/390f157292d7e90f23fde97945e6884ea5f73d88

**Comment by atotic** [comment]
Thanks for doing this, I was just in the middle of merge/rebase :)

Can we rebase to new master before I pull? New master is 4 commits ahead of
old. It'll avoid conflicts in final merge.

One more thing, where should I commit my future changes? New short-lived
branch off GoogleWebComponents/google-map#0.8-preview, that you pull in?

**Comment by atotic** [comment]
One more thing, I've got no write access to g-w-t tree. I can work in atotic

**Comment by atotic** [comment]
And one more: you have strip-trailing-whitespace on in your editor. I
turned mine off to avoid whitespace diffs when I touch the file.

Can I turn it back on? Some offices do not like trailing space diffs.

**Comment by ebidel** [comment]
Should be rebased onto the latest master. We'll stop taking PRs to master to avoid this sort of thing in the future.

Please commit future changes to your cloned `0.8-preview` branch and PR them to this repo's `0.8-preview` branch. Same process as before.

I prefer striping trailing whitespace so feel free to turn it back on.

**Pull request closed without merging**